hoomd/hpmc/UpdaterShape.h

```diff
@@ -0,0 +1,501 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#ifndef _UPDATER_SHAPE_H
+#define _UPDATER_SHAPE_H
+
+#include "IntegratorHPMCMono.h"
+#include "ShapeMoves.h"
+#include "ShapeUtils.h"
+#include "hoomd/HOOMDMPI.h"
+#include "hoomd/Updater.h"
+#include <algorithm>
+#include <numeric>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace hoomd
+    {
+
+namespace hpmc
+    {
+
+template<typename Shape> class UpdaterShape : public Updater
+    {
+    public:
+    UpdaterShape(std::shared_ptr<SystemDefinition> sysdef,
+                 std::shared_ptr<Trigger> trigger,
+                 std::shared_ptr<IntegratorHPMCMono<Shape>> mc,
+                 std::shared_ptr<ShapeMoveBase<Shape>> move);
+
+    ~UpdaterShape();
+
+    virtual void update(uint64_t timestep);
+
+    void initializeDeterminatsInertiaTensor();
+
+    Scalar getTotalParticleVolume()
+        {
+        Scalar volume = 0.0;
+        ArrayHandle<unsigned int> h_ntypes(m_ntypes, access_location::host, access_mode::read);
+        for (unsigned int ndx = 0; ndx < m_ntypes.getNumElements(); ndx++)
+            {
+            detail::MassProperties<Shape> mp(m_mc->getParams()[ndx]);
+            volume += mp.getVolume() * Scalar(h_ntypes.data[ndx]);
+            }
+        return volume;
+        }
+
+    Scalar getShapeMoveEnergy(uint64_t timestep)
+        {
+        Scalar energy = 0.0;
+        ArrayHandle<unsigned int> h_ntypes(m_ntypes, access_location::host, access_mode::readwrite);
+        ArrayHandle<Scalar> h_det(m_determinant_inertia_tensor,
+                                  access_location::host,
+                                  access_mode::readwrite);
+        for (unsigned int ndx = 0; ndx < m_ntypes.getNumElements(); ndx++)
+            {
+            energy += m_move_function->computeEnergy(timestep,
+                                                     h_ntypes.data[ndx],
+                                                     ndx,
+                                                     m_mc->getParams()[ndx],
+                                                     h_det.data[ndx]);
+            }
+        return energy;
+        }
```

Comment:
Why isn't this a function of the move itself?

Reply:
I agree it makes more sense in the shape move classes. If necessary, I can come back to this once I the other more relevant changes are addressed.

Reply:
Yeah, we should prioritize bigger tasks first.

Reply:
This would involve modifying the data storage quite a bit, so we can hold off on this.